src/libraries/System.Private.CoreLib/src/System/IO/LegacyFileStreamStrategy.cs

```diff
@@ -0,0 +1,552 @@
+        /// requests are coordinated with async ones by implementing the sync ones over the async
+        /// ones.  On Unix, we do something similar.  If non-async, async read/write requests just
+        /// delegate to the base stream, and no attempt is made to synchronize.  If async, we use
+        /// a semaphore to coordinate both sync and async operations.
+        /// </summary>
+        private readonly bool _useAsyncIO;
+
+        /// <summary>cached task for read ops that complete synchronously</summary>
+        private Task<int>? _lastSynchronouslyCompletedTask;
+
+        /// <summary>
+        /// Currently cached position in the stream.  This should always mirror the underlying file's actual position,
+        /// and should only ever be out of sync if another stream with access to this same file manipulates it, at which
+        /// point we attempt to error out.
+        /// </summary>
+        private long _filePosition;
+
+        /// <summary>Whether the file stream's handle has been exposed.</summary>
+        private bool _exposedHandle;
+
+        internal LegacyFileStreamStrategy(FileStream fileStream, SafeFileHandle handle, FileAccess access, int bufferSize, bool isAsync) : base(fileStream)
+        {
+            _exposedHandle = true;
+            _bufferLength = bufferSize;
+
+            InitFromHandle(handle, access, isAsync);
+
+            // Note: Cleaner to set the following fields in ValidateAndInitFromHandle,
+            // but we can't as they're readonly.
+            _access = access;
+            _useAsyncIO = isAsync;
+
+            // As the handle was passed in, we must set the handle field at the very end to
+            // avoid the finalizer closing the handle when we throw errors.
+            _fileHandle = handle;
+        }
+
+        internal LegacyFileStreamStrategy(FileStream fileStream, string path, FileMode mode, FileAccess access, FileShare share, int bufferSize, FileOptions options) : base(fileStream)
+        {
+            string fullPath = Path.GetFullPath(path);
+
+            _path = fullPath;
+            _access = access;
+            _bufferLength = bufferSize;
+
+            if ((options & FileOptions.Asynchronous) != 0)
+                _useAsyncIO = true;
+
+            _fileHandle = FileStreamHelpers.OpenHandle(fullPath, mode, access, share, options);
+
+            try
+            {
+                Init(mode, share, path, options);
+            }
+            catch
+            {
+                // If anything goes wrong while setting up the stream, make sure we deterministically dispose
+                // of the opened handle.
+                _fileHandle.Dispose();
+                _fileHandle = null!;
+                throw;
+            }
+        }
+
+        ~LegacyFileStreamStrategy()
+        {
+            // it looks like having this finalizer is mandatory,
+            // as we can not guarantee that the Strategy won't be null in FileStream finalizer
```

Comment:
We _could_ if we wanted to split the ctor from the initialization, e.g.
```C#
_strategy = new Legacy...();
_strategy.Initialize(argumentsPreviouslyPassedToCtor);
```
I don't know whether it's worth it.